nil-docs/zkllvm/getting-started/preparing-input-files.mdx

```diff
@@ -0,0 +1,187 @@
+
+## Specifying keys and values
+
+Note that input files only support the following keys (types).
+
+- `"string"` (C-like strings)
+- `"int"` (integers)
+- `"field"` (rfield elements)
+- `"curve"` (curve elements)
+- `"array"` (`std::array`)
+- `"vector"` (an `ext_vector_type` containing integer, field, or curve elements)
+- `"struct"` (struct or class types)
+
+The table below summarizes what values are accepted by each key.
+
+|                        Key (Type)                         | Values                                                                                               |
+| :-------------------------------------------------------: | :--------------------------------------------------------------------------------------------------- |
+|                        `"string"`                         | Scalar constants                                                                                     |
+| `"int"`, `"field"` and aggregated types that contain them | Scalar constants, strings (in case a number foes not fit within 64 bits), hexadecimal strings        |
+|                         `"curve"`                         | Arrays of numbers containing exactly two values                                                      |
+|             `"array"`, `"vector"`, `"struct"`             | Suitable aggregate types containing exactly the number of elements specified in the type description |
+
+## Extending keys in input files
+
+Except `"int"` and `"string"`, keys, can be extended by specifying additional type details in angle brackets (`"<>"`) as shown in the below examples.
+
+- `"field<pallas_scalar>"`
+- `"vector<int>"`
+
+:::info[Aggregate types]
+
+Aggregate types (`"struct"`, `"array"`, and `"vector"`) are extended with the types of the elements they contain. In case of `"array"` and `"vector"`, only a single type needs to be specified. For `"struct"`, it is possible to specify several types separated by a comma.
+
+:::
+
+:::info[Fields and curves]
+
+The `"field"` and `"struct"` keys can be extended by specifying the kind of their elements.
+
+:::
+
+The `"field"` key supports the following extensions.
+
+- `pallas_base`
+- `bls12381_base`
+- `ed25519_base`
+- `pallas_scalar`
+- `bls12381_scalar`
+- `ed25519_scalar`
+
+The `"curve"` key supports the following extensions.
+
+- `pallas`
+- `bls12381`
+- `ed25519`
+
+If an aggregate type is not extended, its elements must be specified as separate key-value pairs in the input file with keys containing their type descriptions.
+
+Consider the following example.
+
+<Tabs groupId='language'>
+  <TabItem value='cpp' label='C++'>
+
+    ```cpp
+    struct PrivateInput {
+      std::array<int, 3> arr;
+      **int input_number;
+    };
+
+    [[circuit]] void circuit_func([[private_input]] PrivateInput s,
+    const char *str, **std::vector<int, 4> numbers);
+    ```
+
+  </TabItem>
+  <TabItem value='rust' label='Rust'>
+
+    ```rust
```

Comment:
Let's fix this example, because there are a number of things wrong here.

First of all, `std::vector` as well as `Vec` are not currently supported, because they imply dynamic memory allocations.

Second, `"vector"` in input file is not `std::vector` or `Vec`  — it's LLVMs vector, which is a pretty annoying hacky type, used mostly for intrinsics.

Third, arguments to circuit function must be passed by value. It's the same as with `main` function in C++: you can't retrieve pointer or reference from `argv`. So when you have some argument of a circuit function like this:

```cpp
[[circuit]] void f(int *x)
```

that makes no sense because — how will you pass a pointer as input? The special cases, I think, is `const char*`, which refers to a constant string (a global constant in LLVM).

Also I'm not sure, if Rust `&str` in circuit inputs works fine. I'll check this out and come back.

In Rust argument attribute `#[private_input]` is not yet implemented (I just don't have time for that rn). So we probably should either exclude it from Rust example and add in the future, or write it down as it will be and pray that we will implement it soon :)

In Rust example (and probably in C++) I would suggest going with something like this:

```rust
struct PublicInput {
    arg_one: i32,
    arg_two: u32,
}

#[circuit]
fn circuit_func(s: PublicInput, numbers: arr: [i32; 4]) {}
```

And input file then will look like this:

```json
[
  {
    "struct": [ 10, 20 ]
  },
  {
    "array<int>": [ 1, 2, 3, 4 ]
  }
]
```

@makxenov correct me, if I'm wrong somewhere

Reply:
example is fully replaced